### DISCUSSION TITLE
How do you type a function that takes a module as an input?

### DISCUSSION THREAD

**Discussion by Jacob-Flasheye** [python/typing#1408] [Q&A] [answered]
I want to create a function that takes a module as an input:
```python
import foo

def register_module(module: ???):
    ...

register_module(foo)
```
and I've tried looking around but cannot find anything about typing modules as input, probably because it's a uncommon use-case. I'll just on with using `Any` but it would be nice to know if there is a way to type this as well.

**Comment by Gobot1234** [comment, marked as answer]
You are looking for types.ModuleType

**Reply by Jacob-Flasheye** [reply]
Thanks for the quick answer!

**Reply by erictraut** [reply]
`types.ModuleType` won't provide any type checking within the function itself since `types.ModuleType` doesn't define any symbols within the module other than those that are common to all modules such as `__name__` and `__path__`. If your function requires additional symbols to be present in the module, you can define a structural type in the form of a [protocol class](https://peps.python.org/pep-0544/#modules-as-implementations-of-protocols).

**Reply by Jacob-Flasheye** [reply]
Thanks for the follow-up!
I don't want to say I'm a typing newbie (have been using it a lot over the years, love the work!) but some concepts are kind hard to understand. How would I describe a module that has at least one attribute (member?) of a certain class in it?
Like in this example, how should `ModuleWithGenerated` be defined?:
```python
# In generated_1.py
a = _Generated(1, 2, 3)
b = _Generated(4, 5, 6)

# In main.py
from typing import Protocol

import generated_1
import generated_2  # generated_2.py has no attributes of type _Generated

class ModuleWithGenerated(Protocol):
    ... # Must contain at least one _Generated attribute

def register_module(module: ModuleWithGenerated):
    ...

register_module(generated_1)  # OK
register_module(generated_2)  # NOT OK
```

I honestly don't think I _have_ to type this properly because I'm checking the attribute type anyway, but I like to make the typing correct. Gives me a sense of pride and accomplishment!

**Reply by erictraut** [reply]
You can't express the concept "must contain at least one _Generated attribute" with a protocol class, but you can express concepts like "there must be an attribute called `a` whose type is `_Generated`" or "there must be a function called "foo" that has a particular signature.

**Reply by Jacob-Flasheye** [reply]
Gotcha. Suspected as much really, seems like a fairly unusual thing to do.
